File: XPLO/ViewControllers/PhotoAlbumViewController.m
//
//  PhotoAlbumViewController.m
//  XPLO
//
//  Created by Francisco Bernal Yescas on 12/5/17.
//  Copyright © 2017 Sean Fredrick, LLC. All rights reserved.
//

@import CoreMotion;

@import simd;
@import AVFoundation;
@import ImageIO;
@import CoreVideo;
@import Photos;

#import "PhotoAlbumViewController.h"
#import "WMUtilities.h"
#import "WMMatrixUtilities.h"
#import "WMCamera.h"
#import "PreviewView.h"

@interface PhotoAlbumViewController()

@property (nonatomic, assign) BOOL requestPhotoLibrary;
@property (nonatomic, assign) BOOL isPhotoSelected;
@property (nonatomic, strong) NSTimer *updateTimer;
@property (nonatomic, assign) BOOL manual;
@property (nonatomic, assign) float effectRotation;
@property (nonatomic, assign) matrix_float4x4 matrixDeviceOrientation;
@property (weak, nonatomic) IBOutlet PreviewView *previewView;
@property (weak, nonatomic) IBOutlet UIButton *wiggleButton;
@property (weak, nonatomic) IBOutlet UIButton *manualButton;

@end

@implementation PhotoAlbumViewController

static const float kEffectRotationRate = 5.0f;
static const float kEffectRotationRadius = 1.5f;
static const float kEffectMagnificationRate = 0.01f;
static const float kEffectMagnificationMinFactor = 0.90f;
static const float kEffectMagnificationRangeMin = 0.0f;
static const float kEffectMagnificationRangeMax = 30.0f;

- (void)viewDidLoad {
  _isPhotoSelected = NO;
  _requestPhotoLibrary = YES;
  _manual = NO;
  _effectRotation = 0.0f;
  _matrixDeviceOrientation = matrix_identity_float4x4;
  
  [super viewDidLoad];
  
  [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus phstatus) {
    NSLog(@"PHAuthorizationStatus = %d", (int)phstatus);
  }];
  
  // Update timer
  _updateTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 / 60.0
                                                  target:self
                                                selector:@selector(doTimerUpdate)
                                                userInfo:nil
                                                 repeats:YES];
  
  // Gestures
  UITapGestureRecognizer *singleTapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self  action:@selector(didSingleTap:)];
  singleTapRecognizer.numberOfTapsRequired = 1;
  [self.view addGestureRecognizer:singleTapRecognizer];
  
  UITapGestureRecognizer *doubleTapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self  action:@selector(didDoubleTap:)];
  doubleTapRecognizer.numberOfTapsRequired = 2;
  [self.view addGestureRecognizer:doubleTapRecognizer];
  
  UIPinchGestureRecognizer *pinchRecognizer = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(didPinch:)];
  [self.view addGestureRecognizer:pinchRecognizer];
  
  UIPanGestureRecognizer *panRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(didPan:)];
  [panRecognizer setMinimumNumberOfTouches:1];
  [panRecognizer setMaximumNumberOfTouches:1];
  [self.view addGestureRecognizer:panRecognizer];
}

- (void)viewDidAppear:(BOOL)animated {
  if (_requestPhotoLibrary) {
    [self _selectPhotoFromLibrary];
    _requestPhotoLibrary = NO;
  }
}

#pragma mark Selectors

- (IBAction)backButtonTapped:(UIButton *)sender {
  [self dismissViewControllerAnimated:YES completion:NULL];
}

- (IBAction)shareButtonTapped:(UIButton *)sender {
}

- (IBAction)photoAlbumButtonTapped:(UIButton *)sender {
  [self _selectPhotoFromLibrary];
}

- (IBAction)manualButtonTapped:(UIButton *)sender {
  if (sender.isSelected) return;
  sender.selected = YES;
  sender.backgroundColor = [[UIColor whiteColor] colorWithAlphaComponent:0.5];
  _wiggleButton.backgroundColor = [UIColor clearColor];
  _wiggleButton.selected = NO;
  [self useManual];
}

- (IBAction)wiggleButtonTapped:(UIButton *)sender {
  if (sender.isSelected) return;
  sender.selected = YES;
  sender.backgroundColor = [[UIColor whiteColor] colorWithAlphaComponent:0.5];
  _manualButton.backgroundColor = [UIColor clearColor];
  _manualButton.selected = NO;
  [self useTimer];
}

- (void)didSingleTap:(UITapGestureRecognizer*)gestureRecognizer {
  if (!_manual) {
    [self useManual];
  } else {
    [self useTimer];
  }
}

- (void)didDoubleTap:(UITapGestureRecognizer*)gestureRecognizer {
  [self _selectPhotoFromLibrary];
}

- (void)didPinch:(UIPinchGestureRecognizer*)gestureRecognizer {
  static float lastScale = 0.0f;
  
  if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
    lastScale = [gestureRecognizer scale];
  }
  
  if ((gestureRecognizer.state == UIGestureRecognizerStateBegan) ||
      (gestureRecognizer.state == UIGestureRecognizerStateChanged)) {
    const float newScale = (gestureRecognizer.scale - lastScale) * 25.0f;
    
    WMCamera *camera = [_previewView.camera copy];
    const float zPosition = MIN(kEffectMagnificationRangeMax, MAX(kEffectMagnificationRangeMin, [camera zPosition] + newScale));
    const float mag = kEffectMagnificationMinFactor - zPosition * kEffectMagnificationRate;
    
    [camera setZPosition:zPosition];
    [_previewView setCamera:camera];
    _previewView.focalMagnificationFactor = mag;
    
    lastScale = gestureRecognizer.scale;
  }
}

- (void)didPan:(UIPanGestureRecognizer*)gestureRecognizer {
  CGPoint point = [gestureRecognizer translationInView: self.view];
  [gestureRecognizer setTranslation:CGPointZero inView: self.view];
  WMCamera *camera = [_previewView.camera copy];
  
  float pan = (float)point.x / 180.0f * M_PI;
  float tilt = (float)point.y / 180.0f * M_PI;
  
  [camera setPan: camera.pan + pan];
  [camera setTilt: camera.tilt + tilt];
  
  [_previewView setCamera:camera];
}

#pragma mark Animation

-(void)useManual {
  _manual = YES;
  [_updateTimer invalidate];
  _updateTimer = nil;
  _effectRotation = 0.0f;
}

-(void)useTimer {
  _manual = NO;
  _updateTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 / 30.0
                                                  target:self
                                                selector:@selector(doTimerUpdate)
                                                userInfo:nil
                                                 repeats:YES];
}

-(void)doTimerUpdate {
  WMCamera *camera = [_previewView.camera copy];
  
  _effectRotation += kEffectRotationRate;
  _effectRotation = (float)((int)_effectRotation%360);
  
  const float theta = _effectRotation / 180.0f * M_PI;
  const float rx = cosf(-theta) * kEffectRotationRadius;
  const float ry = sinf(-theta) * kEffectRotationRadius;
  
  [camera setXPosition:rx];
  [camera setYPosition:ry];
  [_previewView setCamera:camera];
}

#pragma mark Image Picker Delegate

- (void)imagePickerController:(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary *)info {
  [picker dismissViewControllerAnimated:YES completion:^{ }];
  
  PHAsset *asset = [info objectForKey:UIImagePickerControllerPHAsset];
  
  PHImageRequestOptions *imageRequestOptions = [[PHImageRequestOptions alloc] init];
  imageRequestOptions.synchronous = YES;
  imageRequestOptions.version = PHImageRequestOptionsVersionOriginal;
  imageRequestOptions.networkAccessAllowed = YES;
  
  [[PHImageManager defaultManager]  requestImageDataForAsset:asset
                                                     options:imageRequestOptions
                                               resultHandler:^(NSData *imageData,
                                                               NSString *dataUTI,
                                                               UIImageOrientation orientation,
                                                               NSDictionary *info) {
                                                 NSDictionary *cgImageProperties = [WMUtilities imagePropertiesFromImageData:imageData];
                                                 NSNumber *cgImageOrientationNumber = [cgImageProperties valueForKey:(id)kCGImagePropertyOrientation];
                                                 CGImagePropertyOrientation cgImageOrientation = (cgImageOrientationNumber != nil) ? cgImageOrientationNumber.intValue : kCGImagePropertyOrientationUp;
                                                 const float imageOrientationRadAngle = [WMUtilities radAngleFromImageOrientation:cgImageOrientation];
                                                 
                                                 AVDepthData *depthData = [WMUtilities depthDataFromImageData:imageData];
                                                 if ( !depthData ) {
                                                   if (!_isPhotoSelected) {
                                                     [self dismissViewControllerAnimated:YES completion:NULL];
                                                   }
                                                   return;
                                                 }
                                                 
                                                 [[PHImageManager defaultManager] requestImageForAsset:asset
                                                                                            targetSize:PHImageManagerMaximumSize
                                                                                           contentMode:PHImageContentModeDefault
                                                                                               options:imageRequestOptions
                                                                                         resultHandler:^(UIImage *image, NSDictionary *info) {
                                                                                           AVCameraCalibrationData *cameraCalibrationData = depthData.cameraCalibrationData;
                                                                                           const matrix_float3x3 intrinsicMatrix = matrix_transpose(cameraCalibrationData.intrinsicMatrix); // Metal requires row-major order
                                                                                           const CGSize intrinsicMatrixReferenceDimensions = cameraCalibrationData.intrinsicMatrixReferenceDimensions;
                                                                                           
                                                                                           [_previewView setTextureOrientation:imageOrientationRadAngle];
                                                                                           [_previewView setDepthMapOrientation:-imageOrientationRadAngle];
                                                                                           [_previewView setDepthMap:depthData.depthDataMap
                                                                                                     intrinsicMatrix:intrinsicMatrix
                                                                                  intrinsicMatrixReferenceDimensions:intrinsicMatrixReferenceDimensions];
                                                                                           [_previewView setImageTexture:image];
                                                                                           
                                                                                           _isPhotoSelected = YES;
                                                                                         }];
                                               }];
}

-(void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
  [picker dismissViewControllerAnimated:YES completion:NULL];
  if (!_isPhotoSelected) {
    [self dismissViewControllerAnimated:YES completion:NULL];
  }
}

#pragma mark Private Methods

- (void)_selectPhotoFromLibrary {
  UIImagePickerController *imagePickerController = [[UIImagePickerController alloc] init];
  imagePickerController.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
  imagePickerController.delegate = self;
  [self presentViewController:imagePickerController animated:YES completion:nil];
}

@end

File: XPLO/ViewControllers/PhotoAlbumViewController.swift
//
//  PhotoAlbumViewController.swift
//  XPLO
//
//  Created by Francisco Bernal Yescas on 1/9/18.
//  Copyright © 2018 Sean Fredrick, LLC. All rights reserved.
//

import UIKit
import MetalKit
import Photos

class PhotoAlbumViewController: UIViewController {
  
  @IBOutlet weak var metalView: MTKView!
  @IBOutlet weak var photoAlbumButton: UIButton!
  @IBOutlet weak var wiggleButton: UIButton!
  @IBOutlet weak var collectionView: UICollectionView!
  @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
  
  private var renderer:Renderer!
  private var lastScale: CGFloat = 0
  private var timer:Timer?
  private var angle: Float = 0
  private var images = [PHAsset : UIImage]()
  private var assets = [PHAsset]()
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    renderer = Renderer(withView: metalView)
    
    PHPhotoLibrary.requestAuthorization { (_) in }
    
    let panGestureRecognizer = UIPanGestureRecognizer(target: self,
                                                      action: #selector(PhotoAlbumViewController.panGestureRecognized(_:)))
    view.addGestureRecognizer(panGestureRecognizer)
    
    let pinchGestureRecognizer = UIPinchGestureRecognizer(target: self,
                                                          action: #selector(PhotoAlbumViewController.pinchGestureRecognizer(_:)))
    view.addGestureRecognizer(pinchGestureRecognizer)
    
    DispatchQueue.global().asyncAfter(deadline: .now() + 0.25) {
      self.fetchXploAssets()
    }
    
    wiggleButton.isSelected = true
    wiggleButtonTapped(wiggleButton)
  }
  
  override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
    wiggleButton.isSelected = true
    wiggleButtonTapped(wiggleButton)
  }
  
  override var prefersStatusBarHidden: Bool {
    return true
  }
  
  //MARK: UIGestureRecognizer
  
  @objc func panGestureRecognized(_ panGestureRecognizer: UIPanGestureRecognizer) {
    let velocity = panGestureRecognizer.velocity(in: panGestureRecognizer.view)
    renderer.setRotationVelocity(velocity)
  }
  
  @objc func pinchGestureRecognizer(_ pinchGestureRecognizer: UIPinchGestureRecognizer) {
    if pinchGestureRecognizer.state == .changed {
      let scale = pinchGestureRecognizer.scale - lastScale
      renderer.position.z += Float(scale) * 100
    }
    lastScale = pinchGestureRecognizer.scale
  }
  
  //MARK: Buttons
  
  @IBAction func backButtonTapper(_ sender: UIButton) {
    self.dismiss(animated: true, completion: nil)
  }
  
  @IBAction func photoAlbumButtonTapped(_ sender: UIButton) {
    self.selectPhoto()
  }
  
  @IBAction func wiggleButtonTapped(_ sender: UIButton) {
    sender.isSelected = !sender.isSelected
    sender.backgroundColor = sender.isSelected ? UIColor.white.withAlphaComponent(0.5) : UIColor.clear
    self.timer?.invalidate()
    self.setDefaultOffset()
    if sender.isSelected {
      angle = 0
      timer = Timer.scheduledTimer(timeInterval: 1.0 / 60,
                                   target: self,
                                   selector: #selector(PhotoAlbumViewController.wiggleTimer(_:)),
                                   userInfo: nil,
                                   repeats: true)
    }
  }
  
  //MARK: Timer
  
  @objc private func wiggleTimer(_ timer: Timer) {
    let kEffectRotationRate: Float = 5.0
    let kEffectRotationRadius: Float = 10
    
    angle += kEffectRotationRate;
    angle = Float(Int(angle) % 360)
    
    let theta = angle / 180.0 * .pi;
    let rx = cosf(-theta) * kEffectRotationRadius;
    let ry = sinf(-theta) * kEffectRotationRadius;
    
    renderer.position.x = rx
    renderer.position.y = ry
  }
  
  //MARK: UIImagePickerControllerDelegate
  
  func selectPhoto() {
    let imagePicker = UIImagePickerController()
    imagePicker.sourceType = .photoLibrary
    imagePicker.delegate = self
    self.present(imagePicker, animated: true, completion: nil)
  }
  
  //MARK: Update
  
  func setDefaultOffset() {
    let mesh = self.renderer.mesh
    let offset = -mesh.offset - 80
    self.renderer.setVirtualCameraOffset(offset)
  }
  
  //MARK: assets
  
  func fetchXploAssets() {
    guard let album = PHPhotoLibrary.shared().findAlbum(albumName: kAlbumName) else {
      DispatchQueue.main.async {
        self.activityIndicator.stopAnimating()
        self.selectPhoto()
      }
      return
    }
    
    self.assets.removeAll()
    self.images.removeAll()
    
    let assets = PHAsset.fetchAssets(in: album, options: nil)
    assets.enumerateObjects { (asset, _, _) in
      self.assets.insert(asset, at: 0)
    }
    DispatchQueue.main.async {
      self.activityIndicator.stopAnimating()
      if self.assets.count > 1 {
        self.collectionView.reloadData()
      }
    }
    if let asset = self.assets.first {
      fetchDepth(asset: asset)
    }
  }
  
  func fetchImage(asset: PHAsset,
                  completion: @escaping (UIImage?) -> Void ) {
    if let image = self.images[asset] {
      DispatchQueue.main.async {
        completion(image)
      }
      return
    }
    
    DispatchQueue.global().async {
      let options = PHImageRequestOptions()
      options.isSynchronous = false
      options.version = .original
      options.isNetworkAccessAllowed = true
      let imageManager = PHCachingImageManager()
      let size = CGSize(width: asset.pixelWidth,
                        height: asset.pixelHeight)
      imageManager.requestImage(for: asset,
                                targetSize: size,
                                contentMode: .aspectFill,
                                options: options,
                                resultHandler: { (image, _) in
                                  if let image = image {
                                    self.images[asset] = image
                                  }
                                  DispatchQueue.main.async {
                                    completion(image)
                                  }
      })
    }
  }
  
  func fetchDepth(asset: PHAsset) {
    fetchImage(asset: asset) { (image) in
      guard let image = image else {
        return
      }
      
      let options = PHImageRequestOptions()
      options.isSynchronous = false
      options.version = .original
      options.isNetworkAccessAllowed = true
      
      PHImageManager.default().requestImageData(for: asset,
                                                options: options) { (data, _, _, _) in
                                                  guard let data = data,
                                                    let source = CGImageSourceCreateWithData(data as CFData, nil),
                                                    let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [String : Any],
                                                    let rawValue = properties[kCGImagePropertyOrientation as String] as? UInt32,
                                                    let orientation = CGImagePropertyOrientation(rawValue: rawValue),
                                                    let depthData = AVDepthData(fromSource: source) else {
                                                      return
                                                  }
                                                  
                                                  var radians: Float = 0
                                                  switch orientation {
                                                  case .down:
                                                    radians = .pi
                                                  case .right:
                                                    radians = .pi / 2
                                                  case .left:
                                                    radians = -.pi / 2
                                                  default:
                                                    radians = 0
                                                  }
                                                  
                                                  self.renderer.update(depthData: depthData,
                                                                       image: image,
                                                                       orientation: orientation,
                                                                       radians: radians,
                                                                       mirroring: false,
                                                                       maxDepth: 200)
                                                  self.setDefaultOffset()
      }
    }
  }
  
}

extension PhotoAlbumViewController: UINavigationControllerDelegate, UIImagePickerControllerDelegate {
  
  func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
    picker.dismiss(animated: true, completion: nil)
    if self.assets.count == 0,
      self.renderer.texture == nil {
      self.dismiss(animated: true, completion: nil)
    }
  }
  
  func imagePickerController(_ picker: UIImagePickerController,
                             didFinishPickingMediaWithInfo info: [String : Any]) {
    picker.dismiss(animated: true, completion: nil)
    guard let asset = info[UIImagePickerControllerPHAsset] as? PHAsset else {
      return
    }
    fetchDepth(asset: asset)
  }
  
}

extension PhotoAlbumViewController: UICollectionViewDataSource {
  
  func collectionView(_ collectionView: UICollectionView,
                      numberOfItemsInSection section: Int) -> Int {
    return self.assets.count
  }
  
  func collectionView(_ collectionView: UICollectionView,
                      cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath)
    cell.layer.cornerRadius = 5
    let imageView = cell.viewWithTag(1) as? UIImageView
    self.fetchImage(asset: self.assets[indexPath.row]) { (image) in
      imageView?.image = image
    }
    return cell
  }
  
}

extension PhotoAlbumViewController: UICollectionViewDelegate {
  
  func collectionView(_ collectionView: UICollectionView,
                      didSelectItemAt indexPath: IndexPath) {
    let asset = self.assets[indexPath.row]
    fetchDepth(asset: asset)
  }
  
}
